WIP: Starknet Snap Rough Skeleton (#1509)
MetaMask/metamask-docs: wallet/reference/non-evm-apis/starknet-snap-api.md

  - `casm?`: CompiledContract | `string` - (Optional) - The compiled [casm](https://docs.starknet.io/architecture-and-concepts/smart-contracts/cairo-and-sierra/).
  - `compiledClassHash?`: `string` - (Optional) The compiled class hash from casm.
- `invocationsDetails`: `object` - (Optional) Transaction details object containing:
  - `nonce`: (Optional) Nonce for the transaction.
  - `blockIdentifier`: (Optional) Block identifier for the transaction.
  - `maxFee`: (Optional) Maximum gas fee allowed for the transaction. If not specified, the fee is automatically calculated.
  - `tip`: (Optional) Additional fee to incentivize miners.
  - `paymasterData`: (Optional) Paymaster-related data for the transaction.
  - `accountDeploymentData`: (Optional) Data for account deployment.
  - `nonceDataAvailabilityMode`: (Optional) Mode for nonce data availability.
  - `feeDataAvailabilityMode`: (Optional) Mode for fee data availability.
  - `version`: (Optional) The transaction version.
  - `resourceBounds`: (Optional) The boundaries on resource consumption during the transaction.
  - `skipValidate`: `boolean` - (Optional) Skip validation of the transaction.

### Returns

> The confirmation of sending a transaction on the starknet contract

Reviewer: Can you clarify this? Is the confirmation an object that contains the `transaction_hash` and `class_hash`?

Reviewer 2: Yes, it will returns a object contains both transaction_hash and class_hash

but i also wonder, do we need to mentioned the details response in this API docs

Author: @stanleyyconsensys for consistency with the thee docs, we try to provide both the example and results for the dev. Is it okay if we leave in the results so it matches the others docs?

<TabItem value="Result">

```json
{
  "transaction_hash": "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8"
}
```

</TabItem>
</Tabs>

### `starkNet_extractPublicKey`

Extracts the public key from a Starknet account address.

#### Parameters

- `userAddress`: `string` - Address of the account contract.
- `chainId`: `string` - (Optional) ID of the target Starknet network. The default network is the Starknet Sepolia testnet.

#### Returns

The public key associated with the specified account address (which may differ from the signer's public key).

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_extractPublicKey",
      params: {
        userAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"0x04bfcab3b7ca7e8b3f3b62b2f7f77e9e4b68080bbf8f0f4a1c8f890864d2c7c1d3c45d8b2e3f5f1c27dfeea4c2f5733e90bfc7484e2a690aa9b8ac4559d2e6a8d7"
```

</TabItem>
</Tabs>

### `starkNet_getCurrentNetwork`

> Retrieves the current StarkNet Snap's network.

Reviewer: ```suggestion
Retrieves the current Starknet Snap's network.
```

- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

The list of the scanned user accounts during the recovery process.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_recoverAccounts",
      params: {
        startScanIndex: 0,
        maxScanned: 5,
        maxMissed: 2,
        chainId: "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "address": "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
  "addressIndex": 0,
  "publicKey": "0x04bfcab3b7ca7e8b3f3b62b2f7f77e9e4b68080bbf8f0f4a1c8f890864d2c7c1d3c45d8b2e3f5f1c27dfeea4c2f5733e90bfc7484e2a690aa9b8ac4559d2e6a8d7",
  "addressSalt": "0x789abc123456789abc123456789abc123456789abc123456789abc1234567890",
  "deployTxnHash": "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8",
  "derivationPath": "m/44'/9004'/0'/0/0",
  "chainId": "0x534e5f5345504f4c4941"
}
```

</TabItem>
</Tabs>

### `starkNet_signDeclareTransaction`

> Signs a `DECLARE` transaction using the private key, returns the signature, and uses it to declare the contract class on StarkNet.

Reviewer: ```suggestion
Signs a [`DECLARE` transaction](https://docs.starknet.io/architecture-and-concepts/network-architecture/transactions/#declare-transaction) using the private key, returns the signature, and uses it to declare the contract class on Starknet.
```

#### Parameters

- `address`: `string` - Address of the sender.
- `chainId`: `string` - ID of the target Starknet network. 

> - `details`: `object` - The declare transaction detail to be signed

Reviewer: ```suggestion
- `details`: `object` - The declare transaction details to be signed, including:
```

  - `accountDeploymentData`: (Optional) Data for account deployment.
  - `nonceDataAvailabilityMode`: (Optional) Mode for nonce data availability.
  - `feeDataAvailabilityMode`: (Optional) Mode for fee data availability.
  - `version`: (Optional) The transaction version.
  - `resourceBounds`: (Optional) The boundaries on resource consumption during the transaction.
  - `skipValidate`: `boolean` - (Optional) Skip validation of the transaction.
  - `classHash`: The computed class hash of compiled contract
  - `compiledClassHash`: `string` - (Optional) The compiled class hash from casm.
  - `senderAddress`: `string` - Address of the sender.
  - `chainId`: `string` - ID of the target Starknet network. 

#### Returns

The signature of the transaction that declares a contract class.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_signDeclareTransaction",
      params: {
        senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x534e5f5345504f4c4941",
        details: {

>             version: '0x2',
>             chainId:  "0x534e5f5345504f4c4941",
>             senderAddress: '0xb60e8dd61c5d32be8058bb8eb970870f07233155',
>             classHash: '0x5f3614e8671257aff9ac38e929c74d65b02d460ae966cd826c9f04a7fa8e0d4'

Reviewer: ```suggestion
            version: "0x2",
            chainId:  "0x534e5f5345504f4c4941",
            senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
            classHash: "0x5f3614e8671257aff9ac38e929c74d65b02d460ae966cd826c9f04a7fa8e0d4"
```

        }
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```js
{
  "signature": [
    "0x1b6efab56d39b2acdf5c631c5b1ec1e365e8a36576fa8e3e29ec53566b5782c6", 
    "0x43a9e7f9f9c9238b4d534c5a7621a9c5946f34e72604b539c7bb1ac0a7c2b95b"
  ]
}
```

</TabItem>
</Tabs>

### `starkNet_signMessage`

Signs a typed data message.

#### Parameters

- `typedDataMessage`: `string` - JSON representation of the typed data to be signed.
- `signerAddress`: `string` - Address of the signer.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

The signed hash of typed data.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_signMessage",
      params: {
        typedDataMessage,
        signerAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"1234567890,9876543210"
```

</TabItem>
</Tabs>

> ### starkNet_signTransaction

Reviewer: ```suggestion
### `starkNet_signTransaction`
```

Signs a transaction using the private key and returns the resulting signature.

#### Parameters

- `address`: `string` - Address of the sender.
- `chainId`: `string` - ID of the target Starknet network. 
- `transactions`: `object` - An array of call objects to be executed. Each call contains the target contract address, function name, and call data.

> - `transactionsDetail`: `object` - (Optional) The transaction detail to be signed

Reviewer: ```suggestion
- `transactionsDetail`: `object` - (Optional) The transaction details to be signed, including:
```

  - `nonce`: (Optional) Nonce for the transaction.
  - `blockIdentifier`: (Optional) Block identifier for the transaction.
  - `maxFee`: (Optional) Maximum gas fee allowed for the transaction. If not specified, the fee is automatically calculated.
  - `tip`: (Optional) Additional fee to incentivize miners.
  - `paymasterData`: (Optional) Paymaster-related data for the transaction.
  - `accountDeploymentData`: (Optional) Data for account deployment.
  - `nonceDataAvailabilityMode`: (Optional) Mode for nonce data availability.
  - `feeDataAvailabilityMode`: (Optional) Mode for fee data availability.
  - `version`: (Optional) The transaction version.
  - `resourceBounds`: (Optional) The boundaries on resource consumption during the transaction.
  - `skipValidate`: `boolean` - (Optional) Skip validation of the transaction.

> #### Results

Reviewer: ```suggestion
#### Returns
```

The signature of the transaction.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_signTransaction",
      params: {
        senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x534e5f5345504f4c4941",
        transactions: [
          {
            entrypoint: "transfer",
            calldata: [
              "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4", 
              "1000000000000000000"
            ]
          }
        ]
      }
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
[
  "0x1b6efab56d39b2acdf5c631c5b1ec1e365e8a36576fa8e3e29ec53566b5782c6", 
  "0x43a9e7f9f9c9238b4d534c5a7621a9c5946f34e72604b539c7bb1ac0a7c2b95b"
]
```

</TabItem>
</Tabs>

### `starkNet_switchNetwork`

Switch the current Starknet Snap's network to a different Starknet network.

#### Parameters

- `chainId`: `string` - ID of the target Starknet network. 

> #### Results
> 
> - A boolean flag to indicate if the network switch was successful.

Reviewer: ```suggestion
#### Return

`true` if the network switch was successful, `false` otherwise.
```

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_switchNetwork",
      "params": {
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
true 
```

</TabItem>
</Tabs>

### `starkNet_verifySignedMessage`

Verifies a signed typed data message.

#### Parameters

- `typedDataMessage`: `string` - JSON representation of the original typed data message to be verified.
- `address`: `string` - Address of the signer.
- `signature`: `string` - Signature of the typed data message.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> `true` if the signature is successfully verified; otherwise, it returns `false`.

Reviewer: ```suggestion
`true` if the signature verification was successful, `false` otherwise.
```

(just for consistency)

The examples on this page use the
[`wallet_invokeSnap`](/snaps/reference/wallet-api-for-snaps/#wallet_invokesnap) JSON-RPC method,
which supports all Starknet Snap API methods.

:::note

You can also communicate with the Starknet network using the
[Starknet API](/services/reference/starknet).

:::

## Supported networks

Starknet currently supports two public networks.
Use these networks' chain IDs with the Starknet Snap API methods.

| Network           | Chain ID (Hexadecimal)   |
|-------------------|--------------------------|
| Mainnet           | `0x534e5f4d41494e`       |
| Testnet (Sepolia) | `0x534e5f5345504f4c4941` |

## API methods

### `starkNet_createAccount`

Deploys an account contract.

#### Parameters

- `addressIndex`: `integer` - (Optional) Specific address index of the derived key in
  [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki).
- `deploy`: `boolean` - (Optional) Indicate whether to include send the deploy transaction for the
  account contract.
  The default is `false`.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `gateway/add_transaction` API endpoint.

Reviewer 3: ```suggestion
The address of the account and the transaction hash if the account has been created.
```

  The default is the Starknet Sepolia testnet.

#### Returns

The [status](https://docs.starknet.io/architecture-and-concepts/network-architecture/transaction-life-cycle/)
of the transaction.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_getTransactionStatus",
      params: {
        transactionHash: "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8",
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"ACCEPTED_ON_L2"
```

</TabItem>
</Tabs>

### `starkNet_getValue`

Calls a `VIEW` method on any contract.

#### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `contractFuncName`: `string` - Target function name of the contract.
- `contractCallData`: `string` - (Optional) Call data for the target function with `,` as a separator.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `feeder_gateway/call_contract` API endpoint.

Reviewer 3: The correct reference for starknet API is now : https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json. Not sure where and if we should put this but this is where I am taking the info to update the Returns sections of the API.

```suggestion
The response from Starknet's `starknet_call` API endpoint.
```

- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

The signed hash of typed data.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_signMessage",
      params: {
        typedDataMessage,
        signerAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"1234567890,9876543210"
```

</TabItem>
</Tabs>

### `starkNet_upgradeAccContract`

:::note

This method is supported by the `get-starknet` library.

:::

Upgrades an account contract.

#### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `maxFee`: `string` - (Optional) Maximum gas fee allowed from the sender.
  If not specified, the maximum fee is automatically calculated.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `gateway/call_contract` API endpoint.

Reviewer 3: ```suggestion
The hash of the transaction.
```

---
description: See the Starknet Snap API reference.
sidebar_position: 1
toc_max_heading_level: 2
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Starknet Snap API

With the [Starknet Snap](../../how-to/use-non-evm-networks/starknet/index.md) installed, dapps can
use the Starknet Snap API to interact with users' Starknet accounts (for example, to send transactions).

:::note

> You can also communicate with the Starknet network using the

Reviewer 3: ```suggestion
To read from the Starknet network (get transactions, get account information, ...) you can use the
```

Reviewer 3: We could give some information regarding the chain id. 

The current starknet network has 2 public network. One testnet and the mainnet. The corresponding chain ids are : 

```javascript
SN_MAIN = "0x534e5f4d41494e"
SN_SEPOLIA = "0x534e5f5345504f4c4941"
```

[Starknet API](/services/reference/starknet).
:::

## `starkNet_createAccount`

Deploys an account contract.

### Parameters

- `addressIndex`: `integer` - (Optional) Specific address index of the derived key in
  [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki).
- `deploy`: `boolean` - (Optional) Indicate whether to include send the deploy transaction for the
  account contract.
  The default is `false`.
- `chainId`: `string` - (Optional) ID of the target Starknet network.

>   The default is the Starknet Goerli testnet.

Reviewer 3: ```suggestion
  The default is the Starknet Sepolia testnet.
```

Reviewer 3: Same for all mentions to Goerli

### Returns

The response from Starknet's `gateway/add_transaction` API endpoint.

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_createAccount",
      "params": {
        "addressIndex": 1,
        "deploy": true,

>         "chainId": "0x1"

Reviewer 3: we could provide here an existing chainId, maybe the testnet one ? Or make clear what the chain id should be at the beginning of the page.

      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "transaction_hash": "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8",
  "address": "0xb60e8dd61c5d32be8058bb8eb970870f07233155"
}
```

</TabItem>
</Tabs>

## `starkNet_displayPrivateKey`

Extracts the private key from the deployed Starknet account and displays it in MetaMask.

### Parameters

- `userAddress`: `string` - Address of the account contract.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

Always returns `null` for security reasons.
The private key is only shown in the MetaMask pop-up window.

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_displayPrivateKey",
      params: {
        userAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
null
```

</TabItem>
</Tabs>

> ## `starkNet_estimateAccountDeployFee`

Reviewer 3: needs check

Author: @khanti42 working time

Gets the estimated gas fee for deploying an account contract.

### Parameters

- `addressIndex`: `integer` - (Optional) Specific address index of the derived key in
  [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki).
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

The response from Starknet's `feeder_gateway/estimate_fee` API endpoint.

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_estimateAccountDeployFee",
      params: {
        addressIndex: 0,
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "suggestedMaxFee": "1000000000000000",
  "overallFee": "900000000000000",
  "gasConsumed": "1000000",
  "gasPrice": "1000000000",
  "unit": "wei",
  "includeDeploy": true
}
```

</TabItem>
</Tabs>

> ## `starkNet_estimateFee`

Reviewer 3: needs check

Author: @khanti42 working on this time

- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

The public key of the given account address (can be different from the actual signer).

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_extractPublicKey",
      params: {
        userAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"0x04bfcab3b7ca7e8b3f3b62b2f7f77e9e4b68080bbf8f0f4a1c8f890864d2c7c1d3c45d8b2e3f5f1c27dfeea4c2f5733e90bfc7484e2a690aa9b8ac4559d2e6a8d7"
```

</TabItem>
</Tabs>

## `starkNet_getErc20TokenBalance`

Gets the user's current balance of an ERC-20 token.

### Parameters

- `tokenAddress`: `string` - Address of the ERC-20 token contract.
- `userAddress`: `string` - Address of the user account.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

> The token balance in hexadecimal.

Reviewer 3: It returns an object with the pending and latest balance in hexadecimal : 

e.g.
  ```javascript
{
    balancePending: "0x0",
    balanceLatest: "0x0",
};
```

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_getErc20TokenBalance",
      params: {
        tokenAddress: "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4",
        userAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json

> "0x3e8"

Reviewer 3: ```json
{
  "balancePending": "0x3e8",
  "balanceLatest": "0x3e8",
}
```

```

</TabItem>
</Tabs>

## `starkNet_getStoredUserAccounts`

Gets a list of stored user accounts that are either initialized or initializing.

### Parameters

`chainId`: `string` - (Optional) ID of the target Starknet network.
The default is the Starknet Goerli testnet.

### Returns

> The list of the stored user accounts.

Reviewer 3: There are two new optional return parameters in the account: `upgradeRequired` and `deployRequired`. 

User accounts type is as follow : 

```typescript
export type AccContract = {
  addressSalt: string;
  publicKey: string; // in hex
  address: string; // in hex
  addressIndex: number;
  derivationPath: string;
  deployTxnHash: string; // in hex
  chainId: string; // in hex
  upgradeRequired?: boolean; // whether the account requires an upgrade to Cairo 1
  deployRequired?: boolean; // whether the account requires to be deployed
};
```

Author: @khanti42 I have questions about this one (raised in chat)

Reviewer 2: @khanti42 i think we dont need to mention this method, as it is a method to return the deployed + stored accounts for a single network, consider some edge case for Cairo Contract, it is better not have any starkNet_getStoredXXXX method mentioned in the page

cc @joaniefromtheblock

Author: this has been removed

- `senderAddress`: `string` - Address of the sender.
- `maxFee`: `string` - (Optional) Maximum gas fee allowed from the sender.
  If not specified, the maximum fee is automatically calculated.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

The response from Starknet's `gateway/add_transaction` API endpoint.

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_sendTransaction",
      params: {
        contractAddress: "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4",
        contractFuncName: "transfer",
        contractCallData: "0x456...,0x789...,100",
        senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        maxFee: "1000000000000000",
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "transaction_hash": "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8"
}
```

</TabItem>
</Tabs>

> ## `starkNet_signMessage`

Reviewer 3: needs check

Author: @khanti42 working on this time

        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"1234567890,9876543210"
```

</TabItem>
</Tabs>

## `starkNet_upgradeAccContract`

Upgrades an account contract.

### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `maxFee`: `string` - (Optional) Maximum gas fee allowed from the sender.
  If not specified, the maximum fee is automatically calculated.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Goerli testnet.

### Returns

The response from Starknet's `gateway/call_contract` API endpoint.

### Example

<Tabs>
<TabItem value="Request">

```js
await window.ethereum.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_upgradeAccContract",
      params: {
        contractAddress: "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4",
        maxFee: "1000000000000000",
        chainId: "0x1"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "transaction_hash": "0xdef..."
}
```

</TabItem>
</Tabs>

> ## `starkNet_verifyMessage`

Reviewer 3: needs check

Author: @khanti42 working on this time

  associated with the invocations, such as nonce and version.

#### Returns

A promise that resolves to an `EstimateFeeResponse` object, which contains the following properties:

- `suggestedMaxFee`: `string` - The maximum suggested fee for the transaction, in wei.
  This value is originally a `bigint` and is converted to a base-10 `string`.
- `overallFee`: `string` - The overall fee for the transaction, in wei.
  This value is originally a `bigint` and is converted to a base-10 `string`.
- `unit`: `string` - The unit of the fees, typically `wei`.
- `includeDeploy`: `boolean` - Whether the transaction includes an account deployment step.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_estimateFee",
      params: {
        contractAddress: "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4",
        contractFuncName: "transfer",
        contractCallData: "0x456...,0x789...,100",
        senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",

>         "chainId": "0x534e5f5345504f4c4941"

Reviewer: These example params don't match the params described for `estimateFee`.

The examples on this page use the
[`wallet_invokeSnap`](/snaps/reference/wallet-api-for-snaps/#wallet_invokesnap) JSON-RPC method,
which supports all Starknet Snap API methods.

:::note

You can also communicate with the Starknet network using the
[Starknet API](/services/reference/starknet).

:::

## Supported networks

Starknet currently supports two public networks.
Use these networks' chain IDs with the Starknet Snap API methods.

| Network           | Chain ID (Hexadecimal)   |
|-------------------|--------------------------|
| Mainnet           | `0x534e5f4d41494e`       |
| Testnet (Sepolia) | `0x534e5f5345504f4c4941` |

## API methods

### `starkNet_createAccount`

Deploys an account contract.

#### Parameters

- `addressIndex`: `integer` - (Optional) Specific address index of the derived key in
  [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki).
- `deploy`: `boolean` - (Optional) Indicate whether to include send the deploy transaction for the
  account contract.
  The default is `false`.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `gateway/add_transaction` API endpoint.

Reviewer: Since the `feeder_gateway`/`gateway` has been [deprecated](https://community.starknet.io/t/feeder-gateway-deprecation/100233), should this be updated?

#### Parameters

- `userAddress`: `string` - Address of the account contract.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

Always returns `null` for security reasons.
The private key is only shown in the MetaMask pop-up window.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_displayPrivateKey",
      params: {
        userAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
null
```

</TabItem>
</Tabs>

### `starkNet_estimateAccountDeployFee`

Gets the estimated gas fee for deploying an account contract.

#### Parameters

- `addressIndex`: `integer` - (Optional) Specific address index of the derived key in
  [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki).
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `feeder_gateway/estimate_fee` API endpoint.

Reviewer: Since the `feeder_gateway`/`gateway` has been [deprecated](https://community.starknet.io/t/feeder-gateway-deprecation/100233), should this be updated?

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_estimateAccountDeployFee",
      params: {
        addressIndex: 0,
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
{
  "suggestedMaxFee": "1000000000000000",
  "overallFee": "900000000000000",
  "gasConsumed": "1000000",
  "gasPrice": "1000000000",
  "unit": "wei"
}
```

</TabItem>
</Tabs>

### `starkNet_estimateFee`

Gets the estimated gas fee for calling a method on any contract.

#### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `contractFuncName`: `string` - Target function name of the contract.
- `contractCallData`: `string` - (Optional) Call data for the target function with `,` as a separator.
- `senderAddress`: `string` - Address of the sender.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `feeder_gateway/estimate_fee` API endpoint.

Reviewer: Since the `feeder_gateway`/`gateway` has been [deprecated](https://community.starknet.io/t/feeder-gateway-deprecation/100233), should this be updated?

  The default is the Starknet Sepolia testnet.

#### Returns

The [status](https://docs.starknet.io/architecture-and-concepts/network-architecture/transaction-life-cycle/)
of the transaction.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_getTransactionStatus",
      params: {
        transactionHash: "0x05a56e2d52c817161883f50c441c3228cfe54d9f84b5b5b8b1c8b8e0e6f7e6d8",
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"ACCEPTED_ON_L2"
```

</TabItem>
</Tabs>

### `starkNet_getValue`

Calls a `VIEW` method on any contract.

#### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `contractFuncName`: `string` - Target function name of the contract.
- `contractCallData`: `string` - (Optional) Call data for the target function with `,` as a separator.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `feeder_gateway/call_contract` API endpoint.

Reviewer: Since the `feeder_gateway`/`gateway` has been [deprecated](https://community.starknet.io/t/feeder-gateway-deprecation/100233), should this be updated?

- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

The signed hash of typed data.

#### Example

<Tabs>
<TabItem value="Request">

```js
await getEip6963Provider.request({
  method: "wallet_invokeSnap",
  params: {
    snapId: "npm:@consensys/starknet-snap",
    request: {
      method: "starkNet_signMessage",
      params: {
        typedDataMessage: "{ ... }",
        signerAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155",
        "chainId": "0x534e5f5345504f4c4941"
      },
    },
  },
})
```

</TabItem>
<TabItem value="Result">

```json
"1234567890,9876543210"
```

</TabItem>
</Tabs>

### `starkNet_upgradeAccContract`

:::note

This method is supported by the `get-starknet` library.

:::

Upgrades an account contract.

#### Parameters

- `contractAddress`: `string` - Address of the target contract.
- `maxFee`: `string` - (Optional) Maximum gas fee allowed from the sender.
  If not specified, the maximum fee is automatically calculated.
- `chainId`: `string` - (Optional) ID of the target Starknet network.
  The default is the Starknet Sepolia testnet.

#### Returns

> The response from Starknet's `gateway/call_contract` API endpoint.

Reviewer: Since the `feeder_gateway`/`gateway` has been [deprecated](https://community.starknet.io/t/feeder-gateway-deprecation/100233), should this be updated?

---
description: See the Starknet Snap API reference.
sidebar_position: 1
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Starknet Snap API

When connected to the [Starknet Snap](../../how-to/use-non-evm-networks/starknet/index.md), dapps
can use the Starknet Snap API to interact with users' Starknet accounts (for example, to send transactions).

> Currently, the [`get-starknet`](https://github.com/starknet-io/get-starknet) library only supports the

Reviewer 2: do we really need to specific what SNAP api does get-starknet implemented

as the starknet community (builders) usually following the general standard on client side , which is get-starknet / starknet.js, they have no idea what Starknet Snap API is

i think we can change the angle, instead of telling ppl what API that we supported in get-starknet, we can tell ppl MetaMask Starknet Snap is almost fully support the functionality of get-starknet v3.3.0 (except on/off event listener)

Reviewer 2: @Montoya we have a discussion, what i try to say is using an API to represent a starknet network operation/function is not correct
e.g get-starknet does not integrate starkNet_estimateFee, but it doesnt mean it cant do it, it is still possible to estimate the fee

so me and @khanti42 come up an conclusion that we use operation/method as the category, and each entity (get-starknet, snap), we attach the way to how to perform such action

otherwise, we make the get-starknet only support very few snap api, and seem we only use snap api to interact with the network

Reviewer 4: Ah great, that makes sense

following Starknet Snap API methods:

- [`starkNet_extractPublicKey`](#starknet_extractpublickey)
- [`starkNet_signMessage`](#starknet_signmessage)
- [`starkNet_upgradeAccContract`](#starknet_upgradeacccontract)

> - [`starkNet_verifyMessage`](#starknet_verifymessage)

Reviewer: Does this need to be updated to reflect the latest documentation changes? I.e., does `get-starknet` support a lot of these functionalities according to the new table on the Starknet index page?

Author: updated the docs to call this out
